app/service/rpc_service.ts

```diff
@@ -238,6 +229,102 @@ function lengthPrefixMessage(requestData: Uint8Array) {
   return new Uint8Array(frame);
 }
 
+interface StreamState {
+  buffer: Uint8Array;
+  bufferedLength: number;
+  messageExpectedLength: number;
+  leftovers: Uint8Array;
+}
+
+async function transformLengthPrefixedReaderStreamToProtoMessageBytes(
+  reader: ReadableStreamDefaultReader | undefined,
+  callback: (e: Error | null, b: Uint8Array) => void
+) {
+  let streamState: StreamState = {
+    buffer: new Uint8Array(),
+    bufferedLength: 0,
+    messageExpectedLength: 0,
+    leftovers: new Uint8Array(),
+  };
+  while (true) {
+    // Start with any leftovers from the last turn.
+    let value: Uint8Array | undefined = streamState.leftovers;
+
+    // If we don't have any leftovers, read at least 5 bytes from the stream so we have at least a length prefix.
+    if (streamState.leftovers.length == 0) {
+      value = await readAtLeastNBytes(value, reader, 5);
+    }
+
+    // If we don't have any bytes, the stream is done.
+    if (!value) {
+      return;
+    }
+
+    // If we don't have an expected message length, we can consume the length prefix to get it.
+    if (streamState.messageExpectedLength == 0) {
+      value = consumeLengthPrefix(streamState, value);
+    }
+
+    // Read the rest of the bytes into the buffer until it's full, saving any leftovers.
+    consumeMessageChunk(streamState, value);
+
+    // If we still haven't receive the full message, loop again and read more until we do.
+    if (streamState.bufferedLength < streamState.messageExpectedLength) {
+      continue;
+    }
```

Comment:
sorry I randomly thought about this after approving 😅  should this be

```suggestion
    // Read the rest of the bytes into the buffer until it's full, saving any leftovers.
    while (streamState.bufferedLength < streamState.messageExpectedLength) {
      consumeMessageChunk(streamState, value);
    }
```

or maybe it'd be simpler to use the `readAtLeastNBytes` func to read `bufferedLength - messageExpectedLength` bytes after consuming the length prefix. So that way it's just two calls to `readAtLeastNBytes` on each iteration